6653 contenttype menu (#8032)
plone/volto: docs/how-to-guides/customize-toolbar.md

---
myst:
  html_meta:
    "description": "How to customize the toolbar using pluggables"
    "property=og:description": "How to customize the toolbar using pluggables"
    "property=og:title": "Customize the toolbar"
    "keywords": "Plone, Volto, Seven, Toolbar, Pluggables"
---

# Customize the toolbar

The toolbar uses the pluggable system for extensibility.
It exposes two pluggables:

`toolbar-top`

> :   Rendered at the top of the toolbar. Typically used for action buttons (save, cancel, edit, add).

Reviewer: ```suggestion
:   Rendered at the top of the toolbar.
    Typically used for action buttons, such as save, cancel, edit, and add.
```

`toolbar-bottom`

> :   Rendered at the bottom of the toolbar. Typically used for settings or navigation.

Reviewer: ```suggestion
:   Rendered at the bottom of the toolbar.
    Typically used for settings or navigation.
```

## Register a toolbar button

Wrap your component in a `<Plug>` and place it inside the existing `<PluggablesProvider>` in your layout.

```tsx
import { Plug } from '@plone/layout/components/Pluggable';
import MyIcon from './my-icon.svg?react';
import { Link } from 'react-aria-components';
import { useLocation } from 'react-router';

// Inside your layout component:
const location = useLocation();

<Plug pluggable="toolbar-top" id="my-button">
  <Link href={`/some-action${location.pathname}`} aria-label="My action">
    <MyIcon />
  </Link>
</Plug>
```

> Any data your component needs (content, location, navigation) must be fetched via hooks inside the component itself or captured in the `<Plug>` closure — the toolbar does not pass props down to plugged components.

Reviewer: ```suggestion
Any data your component needs—including content, location, or navigation—must be fetched via hooks inside the component itself or captured in the `<Plug>` closure.
The toolbar does not pass props down to plugged components.
```

## Add a menu to the toolbar

Menus in the toolbar require the `ToolbarMenu` wrapper component from `@plone/layout`.
Use it in place of the standard `Menu` component.
The menu children follow the same API as [React Aria's Menu](https://react-aria.adobe.com/Menu).

> ### Create the menu component

Reviewer: ```suggestion
### Create the menu component

First, create the menu component, using the following example code.
```

```tsx
import { ToolbarMenu } from '@plone/layout/components/Toolbar/ToolbarMenu';
import { MenuItem } from '@plone/components';
import { Header, Menu, MenuSection, Text } from 'react-aria-components';
import type { Content } from '@plone/types';
import MyIcon from './my-icon.svg?react';

import menuStyles from './MyMenu.css?inline';

interface MyMenuProps {
  content: Content;
}

export const MyMenu = ({ content }: MyMenuProps) => {
  return (
    <ToolbarMenu icon={<MyIcon />} styles={menuStyles}>
      <Menu className="menu-my-menu">
        <MenuSection>
          <Header>My menu</Header>
          <MenuItem id="item-1" href="/some-action">
            <Text slot="label">Item 1</Text>
          </MenuItem>
        </MenuSection>
      </Menu>
    </ToolbarMenu>
  );
};
```

`ToolbarMenu` extends `MenuTrigger` and thus `BasicMenuTriggerProps`.
In addition to those props, it accepts the following:

`icon`

> :   The trigger element rendered in the toolbar (typically an SVG icon).

Reviewer: ```suggestion
:   The trigger element rendered in the toolbar, typically an SVG icon.
```

`styles`
:   A CSS string imported with `?inline`.
    Injects the styles into the toolbar's shadow root so that your menu styles apply correctly.

> ### Register the menu as a pluggable

Reviewer: ```suggestion
### Register the menu as a pluggable

Then, register the menu as a pluggable, using the following code.
```

```tsx
import { Plug } from '@plone/layout/components/Pluggable';
import { MyMenu } from './MyMenu';
import { useRouteLoaderData } from 'react-router';
import type { RootLoader } from 'seven/app/root';

// Inside your layout component:
const rootData = useRouteLoaderData<RootLoader>('root');

<Plug pluggable="toolbar-top" id="my-menu">
  <MyMenu content={rootData.content} />
</Plug>
```

Using the same `id` as an existing plug replaces it.

> For a full example, see the `ContentTypesMenu` component in `@plone/publicui`.

Reviewer: This doesn't exist.

https://github.com/search?q=repo%3Aplone%2Fvolto%20ContentTypesMenu&type=code

Reviewer 2: It is introduced with this PR

Author: Yes, as @pnicolli said, it was added in my branch and is located here: https://github.com/plone/volto/blob/6653-contenttype_menu/packages/publicui/components/Toolbar/ContentTypesMenu.tsx

Reviewer: OK, let's link to the future location after merge, and make a note to verify after merge.

```suggestion
For a full example, see the [`ContentTypesMenu`](https://github.com/plone/volto/blob/seven/packages/publicui/components/Toolbar/ContentTypesMenu.tsx) component in `@plone/publicui`.
```

> See the [Pluggables documentation](https://6.docs.plone.org/volto/development/pluggables.html) for the full pluggable API reference.

Reviewer: Don't hard link to Volto documentation, because this link will not get flagged as invalid when Seven is released, nor will it update dynamically.

```suggestion
See the {doc}`plone:volto/development/pluggables` for the full pluggable API reference.
```
